terrawrap/models/graph.py

```diff
@@ -0,0 +1,229 @@
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=num_parallel) as executor:
+            for source in sources:
+                entry = self._get_or_create_entry(source)
+                if not self._has_prefix(entry):
+                    future = executor.submit(entry.no_op)
+                    futures_to_paths[future] = entry.path
+                    continue
+
+                print("Executing %s %s ..." % (entry.path, self.command))
+                future = executor.submit(entry.execute, self.command, debug=debug)
+                futures_to_paths[future] = entry.path
+
+            for future in concurrent.futures.as_completed(futures_to_paths):
+                path = futures_to_paths[future]
+                if self.graph_dict.get(path).state != "no-op":
+                    exit_code, stdout, changes_detected = future.result()
+
+                    if print_only_changes and not changes_detected:
+                        stdout = ["No changes detected.\n"]
+
+                    print("\nFinished executing %s %s ..." % (path, self.command))
+                    print("Output:\n\n%s\n" % "".join(stdout).strip())
+
+                    if exit_code != 0:
+                        self.failures.append(path)
+
+                successors = list(self.graph.successors(path))
+                if successors:
+                    self.recursive_executor(executor, successors, num_parallel, debug, print_only_changes)
+
+        for node in self.graph:
+            item = self.graph_dict.get(node)
+            if not item:
+                self.not_applied.add(node)
+            else:
+                if item.state == "no-op":
+                    self.not_applied.add(item.path)
+        if self.not_applied:
+            print("The following directories have not been run since they are out of scope")
+            print(self.not_applied)
+
+        if self.failures:
+            raise RuntimeError(
+                "The follow directories failed with command '%s':\n%s" % (self.command, "\n".join(self.failures))
+            )
+
+    def recursive_executor(
+            self,
+            executor: concurrent.futures.Executor,
+            successors: List[str],
+            num_parallel: int,
+            debug: bool,
+            print_only_changes: bool):
+        """
+        Helper function for executing graph entries recursively
+        :param executor: The Executor to use. See concurrent.futures.Executor.
+        :param successors: A list of successors to be executed from the previous call
+        :param num_parallel: The number of pipeline entries to run in parallel.
+        :param debug: True if Terraform debugging should be turned on.
+        :param print_only_changes: True if only directories which contained changes should be printed.
+        """
+        futures_to_paths = {}
+
+        for node in successors:
+            entry = self._get_or_create_entry(node)
+            if entry.state is not "Pending":
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Comparison to literal](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)